tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,267 @@
+
+use restate_core::protobuf::node_ctl_svc::{
+    node_ctl_svc_client::NodeCtlSvcClient, GetMetadataRequest,
+};
+use restate_types::{
+    net::AdvertisedAddress,
+    nodes_config::{NodesConfiguration, Role},
+    protobuf::common::{MetadataKind, NodeStatus},
+    storage::{StorageCodec, StorageDecodeError},
+    PlainNodeId, Version,
+};
+
+use crate::util::grpc_channel;
+
+#[derive(Clone, Parser, Collect, Debug)]
+pub struct ConnectionInfo {
+    // todo: rename this to be a node address for reusability across commands
+    /// Cluster Controller address
+    #[clap(
+        long="address",
+        value_hint = clap::ValueHint::Url,
+        default_value = "http://localhost:5122/",
+        env = "RESTATE_ADDRESS",
+        global = true
+    )]
+    pub addresses: Vec<AdvertisedAddress>,
+
+    /// Sync metadata from metadata store first
+    #[arg(long)]
+    pub sync_metadata: bool,
+
+    #[clap(skip)]
+    nodes_configuration: Arc<Mutex<Option<NodesConfiguration>>>,
+
+    #[clap(skip)]
+    cache: Arc<Mutex<HashMap<PlainNodeId, Channel>>>,
+}
+
+impl ConnectionInfo {
+    /// Iterates over all channels associated with the provided addresses
+    ///
+    /// Note: this normally should be private but it's only public because
+    /// it is needed by the `cluster provision` subcommand.
+    pub fn channels(&self) -> impl Iterator<Item = (AdvertisedAddress, Channel)> + use<'_> {
+        self.addresses.iter().cloned().map(|address| {
+            let channel = grpc_channel(address.clone());
+            (address, channel)
+        })
+    }
+
+    /// Gets NodesConfiguration object. This function tries all provided addresses and makes sure
+    /// nodes configuration is cached.
+    pub async fn get_nodes_configuration(&self) -> Result<NodesConfiguration, ConnectionInfoError> {
+        let mut config_guard = self.nodes_configuration.lock().await;
+        if let Some(config) = &*config_guard {
+            return Ok(config.clone());
+        }
+
+        if self.addresses.is_empty() {
+            return Err(ConnectionInfoError::NoAvailableNodes(NoRoleError(None)));
+        }
+
+        let mut latest_configuration: Option<NodesConfiguration> = None;
+        let mut answer = false;
+        let mut errors = NodesErrors::default();
+        let mut cache = self.cache.lock().await;
+
+        for (address, channel) in self.channels() {
+            let mut client = NodeCtlSvcClient::new(channel.clone())
+                .accept_compressed(CompressionEncoding::Gzip)
+                .send_compressed(CompressionEncoding::Gzip);
+
+            let response = match client.get_ident(()).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address, status);
+                    continue;
+                }
+            };
+
+            // node is reachable and answered
+            answer = true;
+            if response.status != NodeStatus::Alive as i32 {
```

Comment:
The response should also contain what is the metadata version the node has. Maybe this is something we can use to get the latest `Logs` or other metadata information we need at some later point.